src/Microsoft.Repl.Tests/ShellTests.cs

```diff
@@ -0,0 +1,191 @@
+        {
+            Shell shell = CreateShell(consoleKey: ConsoleKey.Escape,
+                caretPosition: 0,
+                previousCommand: null,
+                nextCommand: null,
+                out CancellationTokenSource cancellationTokenSource);
+
+            string inputBufferTextBeforeKeyPress = "get";
+            string inputBufferTextAfterKeyPress = string.Empty;
+
+            IShellState shellState = shell.ShellState;
+            shellState.InputManager.SetInput(shellState, inputBufferTextBeforeKeyPress);
+
+            await shell.RunAsync(cancellationTokenSource.Token).ConfigureAwait(false);
+
+            // Verify the input buffer contents after Escape key press event
+            Assert.Equal(inputBufferTextAfterKeyPress, shell.ShellState.InputManager.GetCurrentBuffer());
+        }
+
+        [Fact]
+        public async Task RunAsync_WithInsertKeyPress_FlipsIsOverwriteModeInInputManager()
+        {
+            Shell shell = CreateShell(consoleKey: ConsoleKey.Insert,
+                caretPosition: 0,
+                previousCommand: null,
+                nextCommand: null,
+                out CancellationTokenSource cancellationTokenSource);
+
+            await shell.RunAsync(cancellationTokenSource.Token).ConfigureAwait(false);
+
+            // Verify IsOverwriteMode flag in input manager is set to false after Insert key press event
+            Assert.True(shell.ShellState.InputManager.IsOverwriteMode);
+        }
+
+        [Fact]
+        public async Task RunAsync_WithUnhandledKeyPress_DoesNothing()
+        {
+            Shell shell = CreateShell(consoleKey: ConsoleKey.F1,
+                caretPosition: 0,
+                previousCommand: null,
+                nextCommand: null,
+                out CancellationTokenSource cancellationTokenSource);
+
+            string inputBufferText = "get";
+
+            IShellState shellState = shell.ShellState;
+            shellState.InputManager.SetInput(shellState, inputBufferText);
+
+            await shell.RunAsync(cancellationTokenSource.Token).ConfigureAwait(false);
+
+            // Verify the input buffer contents after F1 key press event
+            Assert.Equal(inputBufferText, shell.ShellState.InputManager.GetCurrentBuffer());
+        }
+
+        private Shell CreateShell(ConsoleKey consoleKey, int caretPosition, string previousCommand, string nextCommand, out CancellationTokenSource cancellationTokenSource)
+        {
+            var defaultCommandDispatcher = DefaultCommandDispatcher.Create(x => { }, new object());
+
+            Mock<IConsoleManager> mockConsoleManager = new Mock<IConsoleManager>();
+            var cts = new CancellationTokenSource();
+            CancellationToken cancellationToken = cts.Token;
+            ConsoleKeyInfo consoleKeyInfo = new ConsoleKeyInfo('\0', consoleKey, false, false, false);
+            mockConsoleManager.Setup(s => s.ReadKey(cancellationToken))
+                .Callback(() => cts.Cancel())
```

Comment:
nit: might be worth putting a comment here to call out that this is used to break the loop so that shell.RunAsync() returns.

Reply:
Sure, added here:
https://github.com/aspnet/HttpRepl/commit/8b3c9233ec033b53939f13dfb0cf4cd302f9ff8d